07-MemoryManagment/memory_managment_module.c

```diff
@@ -0,0 +1,375 @@
+#include <linux/kernel.h>
+#include <linux/cdev.h>
+#include <linux/device.h>
+#include <linux/module.h>
+#include <linux/fs.h>
+#include <linux/init.h>
+#include <linux/types.h>
+#include <linux/errno.h>
+#include <linux/proc_fs.h>
+#include <linux/uaccess.h>
+#include <linux/slab.h>
+#include <linux/kobject.h>
+#include <linux/sysfs.h>
+#include <linux/hrtimer.h>
+#include <linux/ktime.h>
+#include <linux/timekeeping.h>
+#include <linux/math64.h>
+#include <linux/sched/loadavg.h>
+#include <linux/timer.h>
+#include <linux/clocksource.h>
+#include <linux/jiffies.h>
+#include <linux/time.h>
+#include <linux/tick.h>
+#include <linux/fs.h>
+#include <linux/uaccess.h>
+#include <linux/vmalloc.h>
+#include <linux/mm.h>
+
+#define CLASS_NAME	"memory_timing_module"
+#define VMALLOC_TEST "vmalloc_test"
+#define KZALLOC_TEST "kzalloc_test"
+#define KMALLOC_TEST "kmalloc_test"
+#define GET_FREE_PAGES_TEST "get_free_pages_test"
+
+#define BUF_LEN 2048
+
+#define DBG_INFO(x, arg...) \
+pr_info("Test_module_info: function - %s:%i - "x, __func__, __LINE__, ##arg)
+#define DBG_ERROR(x, arg...) \
+pr_err("Test_module_error: function - %s:%i - "x, __func__, __LINE__, ##arg)
+
+static const unsigned long NS = 1e9;
+
+struct memory_module_data {
+	struct class *attr_class;
+	char buf_str[BUF_LEN];
+	struct timespec time;
+	struct timespec prev_time;
+};
+
+static struct memory_module_data *module_data;
+
+static void absolute_time_calc(struct timespec *cur,
+						struct timespec *prev)
+{
+	unsigned long long time_buf = 0;
+	unsigned int time_remainder = 0;
+
+	time_buf = (unsigned long long)(cur->tv_sec) * NS +
+						cur->tv_nsec;
+
+	time_buf -= ((unsigned long long)(prev->tv_sec) * NS +
+						prev->tv_nsec);
+
+	cur->tv_sec = (long)(div_u64_rem(time_buf,
+						NS, &time_remainder));
+
+	cur->tv_nsec = time_remainder;
```

Comment:
This conversation required multiplication and integer division. It is possible to implement the same conversation using only addition and subtract operations.

Reply:
I've deleted this divison. Please see my latest update.